lib/Optimizer/GraphOptimizer/NodeSplitting.cpp

```diff
@@ -0,0 +1,1560 @@
+/**
+ * Copyright (c) Glow Contributors. See CONTRIBUTORS file.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Optimizer/GraphOptimizer/NodeSplitting.h"
+
+#include <algorithm>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+using namespace glow;
+using llvm::cast;
+using llvm::dyn_cast;
+using llvm::isa;
+
+///===---------------------------------------------------------------------===//
+///                                SliceRange
+///===---------------------------------------------------------------------===//
+/// Dimension range representing a contiguous [start, stop] index interval
+/// (both ends are included) along some tensor dimension. The indices are
+/// assumed to be 0 based (C indexing).
+using DimRange = std::pair<dim_t, dim_t>;
+
+/// Dimension paddings representing a virtual padding before/after a given
+/// dimension range.
+using DimPads = std::pair<dim_t, dim_t>;
+
+/// Slice range utility class representing the ranges for all the dimensions
+/// of a slice obtained by extraction from a bigger tensor.
+class SliceRange {
+
+  /// Vector of ranges for all the dimensions of a slice.
+  std::vector<DimRange> ranges_;
+
+public:
+  SliceRange() = default;
+
+  /// Ctor.
+  SliceRange(std::vector<DimRange> ranges) { ranges_ = ranges; }
+
+  /// Ctor.
+  SliceRange(TypeRef type) {
+    for (auto size : type->dims()) {
+      ranges_.emplace_back(0, size - 1);
```

Comment:
I think we should decide now whether I should change the logic from [start, stop] to [start, stop). This is because my memory is fresh and it should be faster to change it now than later. Regarding the subject I don't think there is a best way to do it. I agree some functions look odd when using [start, stop] but on the other hand the logic for creating the slice range maps for Conv2D for example looks much more clean when using [start, stop] than [start, stop).
So I guess the main driving factors should be:
- consistency with other parts of Glow
- choose the most used convention
- etc.

I'll let you decide if you want me to change it. Again if you want me to change it I should do it now and not later since there are some tricky logic involved in the Conv2D which must be changed and it would be easier now.

Reply:
Done. I changed the SliceRange representation from [start, stop] to [start, stop). I also added a new unit test to make sure I did not screw up something while changing this.